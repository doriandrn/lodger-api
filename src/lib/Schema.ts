import { RxJsonSchema, RxJsonSchemaTopLevel } from "rxdb"
import SchemaError from './Error'
import { env } from './defs/env'

enum Errors {
  missingProps = 'Missing properties on schema %%',
  propExists = 'Property "%%" already exists',
  idUndef = 'ID for field %% cannot be undefined',
  invalidField = 'Invalid field %% supplied'
}

/**
 * @interface LodgerSchema
 */
interface LodgerSchema extends RxJsonSchema {
  add (id: string, field: Field): void
}

type LodgerSchemaOptions = {}

const datamodelDir = ['dev', 'test']
  .indexOf(env) > -1 ? '.schemas' : '.'

type SchemaProperties<Interface> = {
  [k in keyof Interface] ?: RxJsonSchemaTopLevel
}

/**s
 *
 *
 * @class Schema
 * @extends {RxJsonSchema}
 * @implements {LodgerSchema}
 */
export default class Schema<Name extends string, Interface> implements RxJsonSchema, LodgerSchema {
  readonly type = 'object'
  readonly version = 0
  readonly properties : SchemaProperties<Interface> = {}
  readonly required: string[] = []
  readonly indexes : string[] = []

  /**
   * Constructs a valid RxJsonSchema out of a Lodger Form Data item
   *
   * @param {LodgerFormCreator} form
   * @param {boolean} [addCommonMethods]
   *
   * @memberof Schema
   * @returns {RxJsonSchema} schema
   */
  constructor (
    readonly name: Name,
    fields ?: { [i: string]: Field },
    options?: LodgerSchemaOptions
  ) {
    if (!fields || !Object.keys(fields).length)
      throw new SchemaError(Errors.missingFields, { name })

    Object.keys(fields).map(f => this.add(fields[f].id || f, fields[f]))

    if (options) {}
  }

  /**
   * Adds properties programatically as
   * we also need to fill in the required array
   *
   * @param {FieldCreator} field
   * @memberof Schema
   */
  add (id: string, field ?: Field) {
    if (!id)
      throw new SchemaError(Errors.idUndef, field)

    if (this.properties[id])
      throw new SchemaError(Errors.propExists, id)

    if (field && !field.rxSchema)
      throw new SchemaError(Errors.invalidField, field)

    const { rxSchema, v, storage, _index } = field

    if (storage !== 'db') return

    const required =  v && v.indexOf('required') > -1
    this.properties[id] = rxSchema || {}

    if (required && this.required.indexOf(id) < 0)
      this.required.push(id)

    if (_index) {
      this.indexes.push(id)
    }
  }

  get ids () {
    return Object.keys(this.properties)
  }
}
