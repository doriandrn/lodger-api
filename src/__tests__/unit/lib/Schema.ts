import Schema, { commonFields } from '~/lib/Schema'
import schema1 from 'fixtures/schemas/withMethods'
import { isRxSchema } from 'rxdb'

const name = 'schema1'

Object.assign(schema1, { name })

describe('Schema', () => {
  let schema: Schema<string, any>

  beforeAll(() => {
    schema = new Schema(schema1)
    console.info('test schema', schema)
    console.info('schema add', schema.add)
  })

  describe('ctor', () => {
    test('is defined', () => {
      expect(schema).toBeDefined()
    })

    test('matches sshot', () => {
      expect(schema).toMatchSnapshot()
    })

    test('is a valid RxSchema', () => {
      expect(isRxSchema(schema)).toBeTruthy()
    })
  })

  describe('.title', () => {
    test('matches given name', () => {
      expect(schema.title).toBe(name)
    })
  })

  describe('.version', () => {
    test('is 0 on first run', () => {
      expect(schema.version).toBe(0)
    })

    test('increases if a new field is added after being constructed', () => {
      schema.add({ id: 'caca' })
      expect(schema.version).toBe(1)
    })
  })

  describe('add()', () => {
    test('adds ok a a new field', () => {

    })
  })

  describe('.indexables - indexable fields', () => {
    test('contains all fields\' ids with index: true', () => {
      expect(schema.indexables).toContainEqual('')
    })

    test('contains keys from commmon methods', () => {
      const cfskeys = commonFields.map(field => field.id)
      expect(Object.keys(schema.properties)).toContain(cfskeys)
    })
  })

  describe('.load', () => {
    test('overwrites form and returns an instance of schema', async () => {
      const sch = await Schema.load('Asociatie')
      expect(typeof sch).toBe(Schema)
    })
  })
})
