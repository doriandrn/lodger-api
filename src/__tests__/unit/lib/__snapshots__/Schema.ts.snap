// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Schema ctor matches sshot: schema1 1`] = `
Schema {
  "indexes": Array [
    "x2",
  ],
  "name": "schema1",
  "properties": Object {
    "x1": Object {
      "type": "string",
    },
    "x2": Object {
      "index": true,
      "type": "string",
    },
    "x3": Object {
      "_type": "number",
      "type": "number",
    },
  },
  "required": Array [
    "x2",
  ],
  "type": "object",
  "version": 0,
}
`;
